### DISCUSSION TITLE
Multiple Python venv in a single quarto website?

### DISCUSSION THREAD

**Discussion by asmaier** [quarto-dev/quarto-cli#1548] [Q&A]
I wanted to create a website with a list of all my quarto presentations. So I created a quarto website and moved all my presentations into subfolders. I include the presentations as iframes into a table in my main website index.qmd. 

```
presentations
│   ├── _quarto.yml
│   ├── _site
│   ├── index.qmd
│   ├── styles.css
│   ├── presentation_1
│   │   ├── Pipfile
│   │   ├── Pipfile.lock
│   │   ├── presentation_1.qmd
│   ├── presentation_2
│   │   ├── Pipfile
│   │   ├── Pipfile.lock
│   │   ├── presentation_2.qmd
...
```

However when rendering the whole website I realized that my presentations use separate virtual Python environments (I use `pipenv`, see the Pipfiles above). So to render my website successfully I have to navigate to each presentation subdirectory, activate the venv for that directory and then render with quarto only the content of the subdirectory. In the end I activate the global environment and render the whole website.

Is there are more convenient way to do that? Is it maybe possible to specify a venv for each qmd file separately? Because I do not want to move all dependencies into my global environment. The dependencies might be conflicting.

**Comment by jcarpi** [comment]
I am also interested in this functionality. Is there a way to do this with "pre-render" scripting?

**Comment by mcanouil** [comment]
If you want to render your presentation first, then render your website, Quarto can't use multiple environment and switch between those because it's a single process.

Now you could write a bash script which takes care of rendering your presentations, _i.e_, going to directory, activating virtual env, rendering, etc., see <https://quarto.org/docs/projects/scripts.html#pre-and-post-render>.
